pennylane/hf/tapering.py

```diff
@@ -0,0 +1,207 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for tapering qubits.
+"""
+# pylint: disable=unnecessary-lambda
+import functools
+import itertools
+
+import autograd.numpy as anp
+import pennylane as qml
+from pennylane import numpy as np
+
+
+def _observable_mult(obs_a, obs_b):
+    r"""Multiply two PennyLane observables together.
+
+    Each observable should be a linear combination of Pauli words, e.g.,
+    :math:`\sum_{k=0}^{N} c_k O_k`, and represented as a PennyLane Hamiltonian.
+
+    Args:
+        obs_a (Hamiltonian): first observable
+        obs_b (Hamiltonian): second observable
+
+    Returns:
+        .Hamiltonian: Observable expressed as a PennyLane Hamiltonian
+
+    **Example**
+
+    >>> c = np.array([0.5, 0.5])
+    >>> obs_a = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliZ(1)])
+    >>> obs_b = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(0) @ qml.PauliZ(1)])
+    >>> print(_observable_mult(obs_a, obs_b))
+      (-0.25j) [Z1]
+    + (-0.25j) [Y0]
+    + ( 0.25j) [Y1]
+    + ((0.25+0j)) [Y0 X1]
+    """
+    o = []
+    c = []
+    for i in range(len(obs_a.terms[0])):
+        for j in range(len(obs_b.terms[0])):
+            op, phase = qml.grouping.pauli_mult_with_phase(obs_a.terms[1][i], obs_b.terms[1][j])
+            o.append(op)
+            c.append(phase * obs_a.terms[0][i] * obs_b.terms[0][j])
+
+    return _simplify(qml.Hamiltonian(qml.math.stack(c), o))
+
+
+def _simplify(h, cutoff=1.0e-12):
+    r"""Add together identical terms in the Hamiltonian.
+
+    The Hamiltonian terms with identical Pauli words are added together and eliminated if the
+    overall coefficient is zero.
+
+    Args:
+        h (Hamiltonian): PennyLane Hamiltonian
+
+    Returns:
+        .Hamiltonian: Simplified PennyLane Hamiltonian
+
+    **Example**
+
+    >>> c = np.array([0.5, 0.5])
+    >>> h = qml.Hamiltonian(c, [qml.PauliX(0) @ qml.PauliY(1), qml.PauliX(0) @ qml.PauliY(1)])
+    >>> print(_simplify(h))
+    (1.0) [X0 Y1]
+    """
+    s = []
+    w = list(range(max(h.wires.tolist()) + 1))
```

Comment:
Do we need to support custom wire labels with this function? If yes, I think we should make the wire map something like `dict(zip(h.wires, range(len(h.wires)+1)))`. What are your thoughts on this?

Reply:
^^ I agree, will the context in which this function is called require mapping custom wire labels? Or is it used in situations where we are initializing the circuits ourselves and the wire labels can be assumed to be ints

Reply:
@obliviateandsurrender Using this `dict(zip(h.wires, range(len(h.wires)+1)))` dos not make trouble when the HF state is tapered? Assume if you have H acting on 0, 1, 3, 4 and not 2.

Reply:
For the molecules I'd tested, it didn't cause any trouble for me. Additionally, you may also have to add mapping labels as an argument in the `qml.grouping.string_to_pauli_word` function that you use in line 201.

Reply:
Added wiremap for all conversions from Pauli string to words and vice versa.